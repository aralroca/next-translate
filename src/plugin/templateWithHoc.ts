import { clearCommentsRgx, overwriteLoadLocales } from './utils'

export default function templateWithHoc(
  code: string,
  {
    skipInitialProps = false,
    typescript = false,
    pageName = '__Page_Next_Translate__',
    hasLoadLocaleFrom = false,
  } = {}
) {
  const tokenToReplace = `__CODE_TOKEN_${Date.now().toString(16)}__`
  const codeWithoutComments = code.replace(clearCommentsRgx, '')

  // Replacing all the possible "export default" (if there are comments
  // can be possible to have more than one)
  let modifiedCode = code.replace(/export +default/g, `const ${pageName} =`)

  // It is necessary to change the name of the page that uses getInitialProps
  // to ours, this way we avoid issues.
  const [, , componentName] =
    codeWithoutComments.match(
      /export +default +(function|class) +([A-Z]\w*)/
    ) || []

  if (componentName) {
    modifiedCode = modifiedCode.replace(
      new RegExp(`\\W${componentName}\\.getInitialProps`, 'g'),
      `${pageName}.getInitialProps`
    )
  }

  let template = `
    import __i18nConfig from '@next-translate-root/i18n'
    import __appWithI18n from 'next-translate/appWithI18n'
    ${tokenToReplace}
    export default __appWithI18n(__Page_Next_Translate__, {
      ...__i18nConfig,
      isLoader: true,
      skipInitialProps: ${skipInitialProps},
      ${overwriteLoadLocales(hasLoadLocaleFrom)}
    });
  `

  if (typescript) template = template.replace(/\n/g, '\n// @ts-ignore\n')

  // Use callback to avoid parsing special patterns specific for .replace()
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_string_as_a_parameter
  return template.replace(tokenToReplace, () => {
    return `\n${modifiedCode}\n`
  })
}
