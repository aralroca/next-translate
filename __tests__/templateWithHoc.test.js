import templateWithHoc from '../src/plugin/templateWithHoc'
import prettier from 'prettier'

function clean(code) {
  return prettier.format(code, { parser: 'typescript' })
}

const tests = [
  {
    describe: 'exporting a variable and with an existing HoC',
    code: `
    import withWrapper from 'somewhere'

    function Page() {
      return <div>Hello world</div>
    }

    const anothervariable = withWrapper(Page);
    const somevariable = anothervariable;
    
    export default somevariable
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
  {
    describe: 'exporting a class with a getInitialProps static outside',
    code: `
    import React from 'react';

    export default class Page extends React.Component {
      render() {
        return <div>Hello world</div>
      }
    }

    Page.getInitialProps = () => ({})
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
  {
    describe:
      'exporting a class with a getInitialProps static outside + one commented before',
    code: `
    import React from 'react';
    
    /*
    export default class FakePage extends React.Component {
      render() {
        return <div>Hello world</div>
      }
    } */

    export default class Page extends React.Component {
      render() {
        return <div>Hello world</div>
      }
    }

    // FakePage.getInitialProps = () => ({})
    /* 
      Page.getInitialProps = () => ({})
    */
    Page.getInitialProps = () => ({})
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
  {
    describe: 'exporting a class with a getInitialProps static',
    code: `
    import React from 'react';

    export default class Page extends React.Component {
      static getInitialProps() {
        return {}
      }

      render() {
        return <div>Hello world</div>
      }
    }
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
  {
    describe:
      'exporting a class with a getInitialProps static | exporting apart',
    code: `
    import React from 'react';

    class Page extends React.Component {
      static getInitialProps() {
        return {}
      }

      render() {
        return <div>Hello world</div>
      }
    }

    export default Page
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
  {
    describe: 'exporting a arrow function with Hoc',
    code: `
    import someHoc from 'whatever'

    export default someHoc(() => {
      return <div>Hello world</div>
    })
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
  {
    describe: 'exporting a arrow function with getInitialProps',
    code: `
    const Page = () => <div>Hello world</div>

    Page.getInitialProps = () => ({})

    export default Page
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
  {
    describe: 'page without "export default" should skip any transformation',
    code: `
    const Page = () => <div>Hello world</div>

    Page.getInitialProps = () => ({})

    export default Page
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
  {
    describe:
      'page with multiple commented "export default" + the correct one should do it right',
    code: `
    const Page = () => <div>Hello world</div>

    Page.getInitialProps = () => ({})

    // export default Page
    /*
      export default Page
    */
    export default Page
  `,
    cases: [{ skipInitialProps: false }, { skipInitialProps: true }],
  },
]

describe('templateWithHoc', () => {
  tests.forEach((d) => {
    describe(d.describe, () => {
      d.cases.forEach(({ expected, debug, ...options }) => {
        const fn = debug ? test.only : test
        const testname = Object.entries(options).map(([k, v]) => `${k}: ${v}`)
        fn(testname.join(' | '), () => {
          expect(clean(templateWithHoc(d.code, options))).toMatchSnapshot()
        })
      })
    })
  })
})
