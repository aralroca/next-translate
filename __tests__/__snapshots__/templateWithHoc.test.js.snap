// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templateWithHoc exporting a arrow function with Hoc skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import someHoc from \\"whatever\\";

const __Page_Next_Translate__ = someHoc(() => {
  return <div>Hello world</div>;
});

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a arrow function with Hoc skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import someHoc from \\"whatever\\";

const __Page_Next_Translate__ = someHoc(() => {
  return <div>Hello world</div>;
});

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a arrow function with getInitialProps skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

const Page = () => <div>Hello world</div>;

Page.getInitialProps = () => ({});

const __Page_Next_Translate__ = Page;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a arrow function with getInitialProps skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

const Page = () => <div>Hello world</div>;

Page.getInitialProps = () => ({});

const __Page_Next_Translate__ = Page;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a class with a getInitialProps static | exporting apart skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

class Page extends React.Component {
  static getInitialProps() {
    return {};
  }

  render() {
    return <div>Hello world</div>;
  }
}

const __Page_Next_Translate__ = Page;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a class with a getInitialProps static | exporting apart skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

class Page extends React.Component {
  static getInitialProps() {
    return {};
  }

  render() {
    return <div>Hello world</div>;
  }
}

const __Page_Next_Translate__ = Page;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a class with a getInitialProps static outside + one commented before skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

/*
    const __Page_Next_Translate__ = class FakePage extends React.Component {
      render() {
        return <div>Hello world</div>
      }
    } */

const __Page_Next_Translate__ = class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
};

// FakePage.getInitialProps = () => ({})
/* 
     __Page_Next_Translate__.getInitialProps = () => ({})
    */
__Page_Next_Translate__.getInitialProps = () => ({});

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a class with a getInitialProps static outside + one commented before skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

/*
    const __Page_Next_Translate__ = class FakePage extends React.Component {
      render() {
        return <div>Hello world</div>
      }
    } */

const __Page_Next_Translate__ = class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
};

// FakePage.getInitialProps = () => ({})
/* 
     __Page_Next_Translate__.getInitialProps = () => ({})
    */
__Page_Next_Translate__.getInitialProps = () => ({});

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a class with a getInitialProps static outside skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

const __Page_Next_Translate__ = class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
};

__Page_Next_Translate__.getInitialProps = () => ({});

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a class with a getInitialProps static outside skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

const __Page_Next_Translate__ = class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
};

__Page_Next_Translate__.getInitialProps = () => ({});

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a class with a getInitialProps static skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

const __Page_Next_Translate__ = class Page extends React.Component {
  static getInitialProps() {
    return {};
  }

  render() {
    return <div>Hello world</div>;
  }
};

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a class with a getInitialProps static skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import React from \\"react\\";

const __Page_Next_Translate__ = class Page extends React.Component {
  static getInitialProps() {
    return {};
  }

  render() {
    return <div>Hello world</div>;
  }
};

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a variable and with an existing HoC skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import withWrapper from \\"somewhere\\";

function Page() {
  return <div>Hello world</div>;
}

const anothervariable = withWrapper(Page);
const somevariable = anothervariable;

const __Page_Next_Translate__ = somevariable;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc exporting a variable and with an existing HoC skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

import withWrapper from \\"somewhere\\";

function Page() {
  return <div>Hello world</div>;
}

const anothervariable = withWrapper(Page);
const somevariable = anothervariable;

const __Page_Next_Translate__ = somevariable;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc page with multiple commented "export default" + the correct one should do it right skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

const Page = () => <div>Hello world</div>;

Page.getInitialProps = () => ({});

// const __Page_Next_Translate__ = Page
/*
      const __Page_Next_Translate__ = Page
    */
const __Page_Next_Translate__ = Page;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc page with multiple commented "export default" + the correct one should do it right skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

const Page = () => <div>Hello world</div>;

Page.getInitialProps = () => ({});

// const __Page_Next_Translate__ = Page
/*
      const __Page_Next_Translate__ = Page
    */
const __Page_Next_Translate__ = Page;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc page without "export default" should skip any transformation skipInitialProps: false 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

const Page = () => <div>Hello world</div>;

Page.getInitialProps = () => ({});

const __Page_Next_Translate__ = Page;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: false,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;

exports[`templateWithHoc page without "export default" should skip any transformation skipInitialProps: true 1`] = `
"import __i18nConfig from \\"@next-translate-root/i18n\\";
import __appWithI18n from \\"next-translate/appWithI18n\\";

(\\"$SomeString$\\");
$$SomeString$$ & $SomeString$ & \`$SomeString$\`;
(\\".cssClass{content:' ';\\");
('Hello \\"world\\"');

const Page = () => <div>Hello world</div>;

Page.getInitialProps = () => ({});

const __Page_Next_Translate__ = Page;

export default __appWithI18n(__Page_Next_Translate__, {
  ...__i18nConfig,
  isLoader: true,
  skipInitialProps: true,
  loadLocaleFrom: (l, n) =>
    import(\`@next-translate-root/locales/\${l}/\${n}\`).then((m) => m.default),
});
"
`;
